bip-taro.mediawiki

```diff
@@ -219,21 +225,37 @@ Taro-specific commitments, then the same or different leaf version can be used
 to gate verification of the new behavior.
 
 An <code>asset_id</code> is the 32-byte hash of: 
-* <code>sha256(genesis_outpoint || asset_tag || asset_meta)</code>
+* <code>sha256(genesis_outpoint || asset_tag || asset_meta || output_position)</code>
 
 where:
 * <code>genesis_outpoint</code> is the first previous input outpoint used in the asset genesis transaction serialized in Bitcoin wire format.
 * <code>asset_tag</code> is a random 32-byte value that represents a given asset, and can be used to link a series of discrete assets into a single asset family. In practice, this will typically be the hash of a human readable asset name.
 * <code>asset_meta</code> is an opaque 32-byte value that can be used to commit to various metadata including external links, documents, stats, attributes, images, etc. Importantly, this field is considered to be ''immutable''.
+* <code>output_position</code> is the output position of the output that caries the unique Taro commitment in the genesis transaction. 
 
 Given the above structure, the <code>asset_id</code> is guaranteed to be
 globally unique from the PoV of the chain, as thanks to
 [[../master/bip-0034.mediawiki|BIP 34]]
 as outpoints can never repeat once serialized within the block chain.
 
 In addition, we enforce a rule that a given output MUST only have a ''single''
-root Taro asset commitment (any other duplicates cannot be spent due to the way
-asset witnesses bind to the input outpoint). 
+root Taro asset commitment. In order to verify this property, we enforce the
+following constraints on the position of the root Taro asset tree commitment:
+* The asset root commitment MUST be the leftmost or rightmost leaf in the tapscript tree:
+** We enforce this by requiring the reveal of the taro commitment in the script tree to either be of depth 0 or 1.
+* For a reveal proof of depth 0, the Taro commitment is the only element in the control block proof, as it directly becomes the root hash.
+* For a reveal proof of depth 1, we require the prover to also reveal the pre-image of the sibling hash.
+** With a root position of depth 1, the Taro commitment is either the leftmost or rightmost element of the fully sorted tapscript tree.
+** Due to the lexicographical sorting of tap branches/leaves before hashing, we lose the ordering information here, and instead force reveal of the sibling pre-iamge.
+** To be a valid Taro root commitment, the sibling hash pre-image MUST either:
+*** Be exactly 32 bytes, which means that the sibling is itself a tap branch (a Taro root pre-image is itself 44 bytes: 2 byte version, 2 byte taro version, 32 byte hash, 8 byte sum value), as bound by its BIP-340 tagged hash tag.
```

Comment:
Isn't the new Taro marker also part of this root preimage?

Reply:
Yes, I added that along the way here as a way to distinguish the pre-image in a unique manner. Related to my comment above, we could even just have an SMT root here, and then use these marker values as keys into that tree, which creates a global commitment space structure/standard in the tapscript tree.